pkg/tun/tun_darwin.go

```diff
@@ -0,0 +1,246 @@
+	if err != nil {
+		return nil, err
+	}
+	return &Device{File: os.NewFile(uintptr(fd), ""), name: name}, nil
+}
+
+// AddSubnet adds a subnet to this TUN device and creates a route for that subnet which
+// is associated with the device (removing the device will automatically remove the route).
+func (t *Device) AddSubnet(_ context.Context, subnet *net.IPNet) error {
+	to := make(net.IP, len(subnet.IP))
+	copy(to, subnet.IP)
+	to[len(to)-1] = 1
+	if err := t.setAddr(subnet, to); err != nil {
+		return err
+	}
+	return withRouteSocket(func(s int) error {
+		return t.routeAdd(s, 1, subnet, to)
+	})
+}
+
+// RemoveSubnet removes a subnet from this TUN device and also removes the route for that subnet which
+// is associated with the device.
+func (t *Device) RemoveSubnet(_ context.Context, subnet *net.IPNet) error {
+	to := make(net.IP, len(subnet.IP))
+	copy(to, subnet.IP)
+	to[len(to)-1] = 1
+	if err := t.removeAddr(subnet, to); err != nil {
+		return err
+	}
+	return withRouteSocket(func(s int) error {
+		return t.routeClear(s, 1, subnet, to)
+	})
+}
+
+func (t *Device) SetMTU(mtu int) error {
+	return withSocket(unix.AF_INET, func(fd int) error {
+		var ifr unix.IfreqMTU
+		copy(ifr.Name[:], t.name)
+		ifr.MTU = int32(mtu)
+		err := unix.IoctlSetIfreqMTU(fd, &ifr)
+		if err != nil {
+			err = fmt.Errorf("set MTU on %s failed: %w", t.name, err)
+		}
+		return err
+	})
+}
+
+// Read reads as many bytes as possible into the given buffer.Data and returns the
+// number of bytes actually read
+func (t *Device) Read(into *buffer.Data) (int, error) {
+	n, err := t.File.Read(into.Raw())
+	if n >= buffer.PrefixLen {
+		n -= buffer.PrefixLen
+	}
+	return n, err
+}
+
+// Write writes bytes from the given buffer.Data and returns the number of bytes
+// actually written.
+func (t *Device) Write(from *buffer.Data) (int, error) {
+	raw := from.Raw()
+	if len(raw) <= buffer.PrefixLen {
+		return 0, unix.EIO
+	}
+
+	ipVer := raw[buffer.PrefixLen] >> 4
+	switch ipVer {
+	case ipv4.Version:
+		raw[3] = unix.AF_INET
+	case ipv6.Version:
+		raw[3] = unix.AF_INET6
+	default:
+		return 0, errors.New("unable to determine IP version from packet")
+	}
+	n, err := t.File.Write(raw)
+	return n - buffer.PrefixLen, err
+}
+
+// Address structure for the SIOCAIFADDR ioctlHandle request
+//
+// See https://www.unix.com/man-page/osx/4/netintro/
+type addrIfReq struct {
+	name [unix.IFNAMSIZ]byte
+	addr unix.RawSockaddrInet4
+	dest unix.RawSockaddrInet4
+	mask unix.RawSockaddrInet4
+}
+
+// Address structure for the SIOCAIFADDR_IN6 ioctlHandle request
+//
+// See https://www.unix.com/man-page/osx/4/netintro/
+type addrIfReq6 struct {
+	name           [unix.IFNAMSIZ]byte
+	addr           unix.RawSockaddrInet6
+	dest           unix.RawSockaddrInet6
+	mask           unix.RawSockaddrInet6
+	flags          int32 // nolint:structcheck
+	expire         int64 // nolint:structcheck
+	preferred      int64 // nolint:structcheck
```

Comment:
Assuming that `time_t` is 32 bits makes me hella nervous, especially when there's `unix.Time_t` to make this easy to get right.

Reply:
... and for some reason it's not defined on GOOS=darwin